pkg/volume/local/local.go

```diff
@@ -205,6 +205,144 @@ func (plugin *localVolumePlugin) ConstructBlockVolumeSpec(podUID types.UID, volu
 	return volume.NewSpecFromPersistentVolume(localVolume, false), nil
 }
 
+var _ volume.DeviceMountableVolumePlugin = &localVolumePlugin{}
+
+type deviceMounter struct {
+	plugin  *localVolumePlugin
+	mounter *mount.SafeFormatAndMount
+}
+
+var _ volume.DeviceMounter = &deviceMounter{}
+
+func (plugin *localVolumePlugin) NewDeviceMounter() (volume.DeviceMounter, error) {
+	return &deviceMounter{
+		plugin:  plugin,
+		mounter: util.NewSafeFormatAndMountFromHost(plugin.GetPluginName(), plugin.host),
+	}, nil
+}
+
+func (dm *deviceMounter) mountLocalBlockDevice(spec *volume.Spec, devicePath string) (string, string, error) {
+	deviceMountPath, err := dm.GetDeviceMountPath(spec)
+	if err != nil {
+		return devicePath, deviceMountPath, err
+	}
+	glog.V(4).Infof("local: mounting device %s to %s", devicePath, deviceMountPath)
+	notMnt, err := dm.mounter.IsLikelyNotMountPoint(deviceMountPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(deviceMountPath, 0750); err != nil {
+				return devicePath, deviceMountPath, err
+			}
+			notMnt = true
+		} else {
+			return devicePath, deviceMountPath, err
+		}
+	}
+	if !notMnt {
+		return devicePath, deviceMountPath, nil
+	}
+	fstype, err := getVolumeSourceFSType(spec)
+	if err != nil {
+		return devicePath, deviceMountPath, err
+	}
+	ro, err := getVolumeSourceReadOnly(spec)
+	if err != nil {
+		return devicePath, deviceMountPath, err
+	}
+	options := []string{}
+	if ro {
+		options = append(options, "ro")
+	}
+	mountOptions := util.MountOptionFromSpec(spec, options...)
+	err = dm.mounter.FormatAndMount(devicePath, deviceMountPath, fstype, mountOptions)
+	if err != nil {
+		os.Remove(deviceMountPath)
+		return devicePath, deviceMountPath, fmt.Errorf("local: failed to mount device %s at %s (fstype: %s), error %v", devicePath, deviceMountPath, fstype, err)
+	}
+	glog.V(3).Infof("local: successfully mount device %s at %s (fstype: %s)", devicePath, deviceMountPath, fstype)
+	return devicePath, deviceMountPath, nil
+}
+
+func (dm *deviceMounter) MountDevice(spec *volume.Spec, devicePath string, _ *v1.Pod) (string, string, error) {
+	if spec.PersistentVolume.Spec.Local == nil || len(spec.PersistentVolume.Spec.Local.Path) == 0 {
+		return devicePath, "", fmt.Errorf("local volume source is nil or local path is not set")
+	}
+	fileType, err := dm.mounter.GetFileType(spec.PersistentVolume.Spec.Local.Path)
+	if err != nil {
+		return devicePath, "", err
+	}
+
+	switch fileType {
+	case mount.FileTypeBlockDev:
+		return dm.mountLocalBlockDevice(spec, devicePath)
```

Comment:
Does `SetUp` also have to change, to bind mount from this global path instead?

Reply:
yes. i think so